### PR TITLE
Fix 'OpenSSL::Digest.{digest,hexdigest,base64digest} algorithm, data' 

### PR DESCRIPTION
**Pull request by bdewater**
I ran into `digest` not working while testing [webauthn-ruby](https://github.com/cedarcode/webauthn-ruby) at `9b08086b`. The test suite failed with examples such as:

```
  1) WebAuthn::AuthenticatorAttestationResponse when TPM attestation verifies
     Failure/Error: OpenSSL::Digest.digest(cose_algorithm.hash_function, att_to_be_signed)
     
     ArgumentError:
       wrong number of arguments (given 2, expected 1)
     # /Users/chrisseaton/src/github.com/Shopify/graal-shopify/sdk/mxbuild/darwin-amd64/GRAALVM_UNKNOWN_JAVA8_STAGE1/graalvm-unknown-java8-20.0.0-dev/Contents/Home/jre/languages/ruby/lib/truffle/digest.rb:139:in `digest'
     # /Users/chrisseaton/src/github.com/Shopify/graal-shopify/sdk/mxbuild/darwin-amd64/GRAALVM_UNKNOWN_JAVA8_STAGE1/graalvm-unknown-java8-20.0.0-dev/Contents/Home/jre/languages/ruby/lib/mri/openssl/digest.rb:35:in `digest'
     # ./lib/webauthn/attestation_statement/tpm.rb:27:in `valid?'
     # ./lib/webauthn/attestation.rb:35:in `valid_attestation_statement?'
```

While fixing this particular error I also noticed `hexdigest` and `base64digest` not working as expected, so I fixed these as well. More details are in the commit messages.

https://github.com/Shopify/truffleruby/issues/1

### CODE REVIEW

File: lib/truffle/digest.rb [hunk 1]
```diff
@@ -136,17 +136,23 @@ def inspect
...
       digest.digest
     end
 
-    def self.hexdigest(message)
-      digest = new
+    def self.hexdigest(*parameters, message)
+      digest = new(*parameters)
       digest.update message
       digest.hexdigest
     end
```
Comment (eregon) [inline lib/truffle/digest.rb]: `Digest.hexdigest` seems defined as `Digest.hexencode(Digest::Class.digest(*args))` in MRI:
https://github.com/ruby/ruby/blob/37c2cd3fa47c709570e22ec4dac723ca211f423a/ext/digest/digest.c#L478-L490

And then https://github.com/oracle/truffleruby/blob/59b78fdbebd53a5cf9d0a4306d26714ffab23e81/lib/mri/openssl/digest.rb#L34-L36 is what, very confusingly, swaps the arguments.

I'll add a commit to do that.

Comment (bdewater) [reply lib/truffle/digest.rb]: Yes this definitely was a brainbender 😅 also had a fun time with how `Digest::Class` and `Digest::Instance` are used depending on using the stdlib Digest or the OpenSSL algorithm modules:

```ruby
OpenSSL::Digest::SHA256.ancestors
=> [OpenSSL::Digest::SHA256, OpenSSL::Digest, Digest::Class, Digest::Instance, Object, Kernel, BasicObject]

Digest::SHA256.ancestors
=> [Digest::SHA256, Digest::Base, Digest::Class, Digest::Instance, Object, Kernel, BasicObject]

Digest::SHA256.ancestors - OpenSSL::Digest::SHA256.ancestors
=> [Digest::SHA256, Digest::Base]
```

File: lib/truffle/digest.rb [hunk 2]
```diff
@@ -136,17 +136,23 @@ def inspect
...
+
+    def self.base64digest(*parameters, message)
+      digest = new(*parameters)
+      digest.update message
+      digest.base64digest
+    end
```
Comment (eregon) [inline lib/truffle/digest.rb]: Just a note: this was already defined at line 29 above:
https://github.com/oracle/truffleruby/pull/1889/files#diff-048c3691437021dc9c9997d1a3aa7792R29

This file is pretty hard to follow due to reopening classes and part of the code from MRI.
Maybe I should try to convince MRI devs to define more of Digest in Ruby :)

**Comment by eregon** [comment]
Merged in 619ad90f5dd3c2a1a67a7b88a509fcf7600c2d74, thank you!
